### PR TITLE
[2.0] Add dotnet-store and runtime-store topics

### LINKED ISSUE #1855: Write documentation for the runtime package cache feature
See core-proposal CR 42

### LINKED ISSUE #2139: [2.0] Create CLI command preview topics
Follow the repo TOC structure when creating the topics in the *preview* folder.

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #1855

### CODE REVIEW

**Comment by GuardRex** [comment]
@mairaw Ready for :eyes:.

Note that the main push here is for the `dotnet store` and `runtime-package-store` topics. There are a few touches here to `dotnet publish`, but the other (broader) `dotnet publish` updates are coming up in https://github.com/dotnet/docs/pull/2480 after these other update PR's are merged.

WRT the file system layout at the top of this topic ...

capture4

I did this on at least one other topic, where I'm trying to use a less "muddy" approach to laying out the file structure. The use of underscores, pipes, and slashes (/, \\) seems messy imho and putting these file layouts into code blocks seems like an ugly hack (again, imho). However (of course), I'll change it on review if you want to go a different way. Maybe this is ok? ... perhaps it just needs a little more indentation (4-spaces instead of 2).

**Comment by GuardRex** [comment]
... and side note: I composed the runtime-package-store topic thinking that they'll make a few minor adjustments to *artifacts.xml* via https://github.com/dotnet/cli/issues/7010. However, they might say that putting the package names into human-friendly casing isn't going to happen. I put the file in here how *I think* they're going to do it, but I'll adjust accordingly based on the dev team's response to that issue.

File: docs/core/tools/dotnet-publish.md
Comment (mairaw) [inline docs/core/tools/dotnet-publish.md]: if we're going to add this, should we show syntax for 1.0 vs 2.0? If not, I'd probably wait to add this only when we get the tabs later this month.

File: docs/core/tools/dotnet-store.md
Comment (mairaw) [inline docs/core/tools/dotnet-store.md]: with the warning, I'd remove this. same for the other topic.

**Review by mairaw — CHANGES_REQUESTED** [review]

**Comment by mairaw** [comment]
The files also need to be added to the TOC

File: docs/core/tools/index.md
Comment (mairaw) [inline docs/core/tools/index.md]: maybe add a note that is for .net core 2.0 only here?

File: docs/core/tools/dotnet-publish.md
Comment (GuardRex) [inline docs/core/tools/dotnet-publish.md]: There is no syntax for 1.0; so in the paragraph below, I put ...

> This option is available starting with .NET Core SDK 2.0 Preview 2.

If I remove it, then we should open an issue to add it back when 2.0 ships. Let me know which way to go.

RE: Tabs -- I'm not sure how the tabs work tho when the prior version doesn't have the feature at all. I guess the 1.0 tab can say something to the effect, "This option isn't available in 1.0."

File: docs/core/tools/dotnet-store.md
Comment (GuardRex) [inline docs/core/tools/dotnet-store.md]: Remove it from the title and H1 .... but leave the includes warning? That's what I'll do, but let me know if you mean *everywhere*.

**Comment by GuardRex** [comment]
@mairaw Couple of questions:

* When you said 'remove preview warning,' I took it off of the title and H1 but left the includes in place. Did you mean *remove it everywhere* or just as I did from the title and H1?
* The `--manifest` option on `dotnet publish`: I have a line in the paragraph saying its available with preview bits. If we kill it for now, then we'll need to open *another issue* to track adding it back. Let me know tho if you want it 🔪 out of there.

File: docs/core/deploying/runtime-package-store.md
Comment (svick) [inline docs/core/deploying/runtime-package-store.md]: Should all the instances of "ASP.NET" here be "ASP.NET Core"?

Comment (svick) [inline docs/core/deploying/runtime-package-store.md]: > the setup packages authored by Microsoft.

Maybe clarify what setup packages is this talking about?

> the .NET Core setup packages authored by Microsoft.

**Comment by svick** [comment]
Should the runtime-package-store article also mention that all this only applies to framework-dependent deployments?

File: docs/core/deploying/runtime-package-store.md
Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: Did you create a store? It should appear at *C:\Users\Svick\.dotnet\store* after you create one. Here's a screenshot of mine after working with it ...

capture

Yes, your confusion is warranted. :cry: This is a blown explanation pulled from another source that was referring either to the store located at *C:\Program Files\dotnet\store* or they actually *MOVED* where a custom package store is created. Either way, I'll change this. Thanks for catching it.

Comment (svick) [inline docs/core/deploying/runtime-package-store.md]: I did not create my own store, I was talking about the one that's created when installing .Net Core. I guess there is a global one in `Program Files\dotnet` and also a user-local one in `Users\Svick\.dotnet` (if you create it)? Not sure how much of this should be explained here.

Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: Yes, that's correct. I fixed that line. The main thrust of the topic is the user-local one, and I don't want to confuse anyone by mentioning the presence of the installed store at this point in the topic. WRT the original text of this topic, they may have moved it at some point ... hard to say.

I think my fix tho is to not discuss the less-relevant store at all until the section *ASP.NET Core implicit store* further down the topic. Even then, I don't want to get too much into the weeds on it. I think your other point about SCD is *very* important to get into that section, and I'm adding that now.

Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: I've been looking at this carefully. Let's *DO* try a version where this isn't mentioned at this point in the topic. I want to keep all of this together in *ASP.NET Core implicit store* if possible (at least until we change our minds about it! :grinning:) Let's take a :eyes: on the next commit.

Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: The content that refers to the installed package store is separate from the content dealing with ASP.NET Core apps. I'll see if I can clarify in the next commit.

**Comment by GuardRex** [comment]
@svick Hopefully, this is at least headed in the right direction.

* I do postpone discussion of the ASP.NET Core target manifests until the topic gets down to that section.
* I try to convey that when the ASP.NET Core target manifests are used (as they come in with `Microsoft.NET.Sdk.Web`), then the required packages better be on the target system when the app gets there or they'll have a nice :boom: :cry:!
* I add a bit to say that you can't use the ASP.NET Core target manifests if its an SCD app. It will throw an error on publish if one tries to set **\<PublishWithAspNetCoreTargetManifest>** to `true`.
* As for the package store that's part of the SDK, I try not to talk about it much on purpose here. Too many package stores described in detail may spoil the stew 🍲, and *two stores* described in any detail here is probably *one too many.* I don't see anything great to link either, but I'm open to suggestions.

File: docs/core/deploying/runtime-package-store.md
Comment (Rick-Anderson) [inline docs/core/deploying/runtime-package-store.md]: See 3:10 into [this video](https://www.dropbox.com/s/a8yscvf4uep5qx8/Metapackage%20%2C%20%20runtime%20store%20and%20publ.%20.%20.%20-%20Thursday%2C%20July%2027%2C%202017%202.33.50%20PM.mp4?dl=0)
The runtime store can be added by installers other than the .NET Core SDK. For example: 
* [apt-get](https://help.ubuntu.com/community/AptGet/Howto)
* [redhat yum](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Deployment_Guide/sec-Installing.html) installer
* windows hosting bundle
* downloading zips
* downloadable if you want to install it manually

Comment (Rick-Anderson) [inline docs/core/deploying/runtime-package-store.md]: The implicit store is added during publishing.  4:40 into video
The all package contains the target manifest for the runtime store.  When you reference `Microsoft.AspNetCore.All`, publishing automatically includes the runtime store manifest as the publish target manifest. Publishing excludes all packages listed in the manifest. 

When you get to trimming, mention all trimming does is exclude packages listed in the manifest.

Comment (Rick-Anderson) [inline docs/core/deploying/runtime-package-store.md]: target manifest is talked about at 5:35

Comment (Rick-Anderson) [inline docs/core/deploying/runtime-package-store.md]: 5:45 - If you reference `Microsoft.AspNetCore.All`, publishing specifies the manifest automatically. Set flag to false to turn off. When set to false, publishing does not trim packages.

**Review by Rick-Anderson — APPROVED** [review]
suggestions

**Comment by Rick-Anderson** [comment]
@GuardRex  see updates in https://github.com/dotnet/designs/issues/8

File: docs/core/deploying/runtime-package-store.md
Comment (mairaw) [inline docs/core/deploying/runtime-package-store.md]: that doesn't take away he's authorship @Rick-Anderson. I'm listed as the ms.author

**Comment by mairaw** [comment]
@GuardRex can you address Rick's comments? I'll ask folks on that other thread to review the topics on staging as well.

File: docs/core/tools/dotnet-store.md
Comment (mairaw) [inline docs/core/tools/dotnet-store.md]: seeing how this topic renders on staging I'm wondering if we should start formatting the options in bold similar to azure cli reference: https://docs.microsoft.com/en-us/cli/azure/account 

The inline code kind of gets mixed up with the options:
image

**Comment by GuardRex** [comment]
@mairaw @Rick-Anderson Yes ... this weekend I'll get it fixed up.

**Comment by mairaw** [comment]
@DamianEdwards do you want to take a look at this too?
Main new topics on staging are:
https://review.docs.microsoft.com/en-us/dotnet/core/tools/dotnet-store?branch=pr-en-us-2496
https://review.docs.microsoft.com/en-us/dotnet/core/deploying/runtime-package-store?branch=pr-en-us-2496

**Comment by Rick-Anderson** [comment]
I think we should add the following information:

The store directory is intended to contain all runtime package store files. Currently,  only ASP.NET core stores packages in the runtime store.

@DamianEdwards  is that OK? Otherwises it's confusing.

**Comment by GuardRex** [comment]
Update: I was busy this weekend working on the custom templates topics and didn't get to this. I plan to hit this on Monday morning. I'll ping here when it's ready for :eyes:.

File: docs/core/deploying/runtime-package-store.md
Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: I don't think I should say "C:\Program Files\ ...." without macOS/Linux info. I'll remove the "user's profile" bit for sure. I'll toss in a */usr/local/share/dotnet/store* for our friends.

btw- Are we standardizing path segment separators as forward-slashes on Windows?

File: docs/core/deploying/runtime-package-store.md
Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: Wait!!! ... We don't get paid *by the word* here??!!! :grinning:

File: docs/core/tools/dotnet-store.md
Comment (GuardRex) [inline docs/core/tools/dotnet-store.md]: @mairaw Sounds good to me, but isn't that a separate issue for discussion with sample renderings? Would u like me to open an issue for it?

File: docs/core/tools/dotnet-publish.md
Comment (GuardRex) [inline docs/core/tools/dotnet-publish.md]: @mairaw I'll leave this for the next author or help from the team. I can't fix this in the time that remains.

File: docs/core/deploying/runtime-package-store.md
Comment (GuardRex) [inline docs/core/deploying/runtime-package-store.md]: I think @DamianEdwards slightly misspoke when he said the runtime store is added during publishing (4:40). Note how he restates it with "infer" immediately after that remark ... i.e., the packages of the runtime store are present (implicitly) when the app is published with the manifest (5:02-5:12) ... and in particular when the `All` metapackage is referenced by the app.

However ... the original language used in this topic is also incorrect (or perhaps just incomplete) imo ...

> The targets in [`Microsoft.NET.Sdk.Web`](https://github.com/aspnet/websdk) include manifests referencing the implicit package store that's available when the .NET Core SDK is installed on the target system.

Those targets may or may not rely on certain packages being present in the runtime store, but it isn't possible that they *require the runtime store implicitly*, which is inferred by the passage ... otherwise SCD deployments simply wouldn't work at all. I'm going to mod the language, and @DamianEdwards can say (as soon as I get the next commit posted) where the language needs additional work. Stand-by for another version of this topic ... I'm about to shoot it in now.

**Comment by tdykstra** [comment]
@GuardRex What is this PR waiting for?

**Comment by guardrex** [comment]
[@]DamianEdwards I guess. 🤷‍♂️

**Comment by tdykstra** [comment]
@guardrex the title in the doc and in the intro and in the toc is still Runtime Store  -- was a decision made to omit Package?

**Comment by guardrex** [comment]
I'll TAL and see ... yes, it was supposed to be removed. I'll double-check it.

**Comment by guardrex** [comment]
Yep ... It's all good. It's "runtime store" everywhere AFAICT. https://github.com/dotnet/docs/pull/2496#discussion_r129976576

**Comment by Rick-Anderson** [comment]
>@guardrex the title in the doc and in the intro and in the toc is still Runtime Store -- was a decision made to omit Package?

The installer uses Runtime Package Store. I think it might be safer to be consistent with the installer.  See https://github.com/dotnet/cli/issues/7301

@richlander  @DamianEdwards  Should we be consistent now?

**Comment by mairaw** [comment]
I think both are out @Rick-Anderson. We might have to check with them later and make a decision ourselves. I'll review it now to see if I see any other issues.

**Comment by mairaw** [comment]
I just left the name of the file runtime store because we'll need redirects to change later. Made some edits. As soon as this finishes building I'll merge, unless someone else has something else to add.

**Comment by mairaw** [comment]
And when you ask, what is the PR waiting for? The answer is usually me. LOL 😄

**Comment by guardrex** [comment]
It would help if *someone* didn't turn in *MASSIVE* giant PR's that need *A LOT* of work. lol

File: docs/core/tools/dotnet-publish.md
Comment (mairaw) [inline docs/core/tools/dotnet-publish.md]: Yes, I'll be doing that. Thanks @guardrex!

**Comment by mairaw** [comment]
We're almost there. Just a couple left now to be reviewed and merged.

**Review by mairaw — APPROVED** [review]

**Pull request merged**
